src/renderers/webgl/WebGLPrograms.js

```diff
@@ -349,6 +326,160 @@ function WebGLPrograms( renderer, cubemaps, cubeuvmaps, extensions, capabilities
+		array.push( parameters.combine );
+		array.push( parameters.vertexUvs );
+		array.push( parameters.fogExp2 );
+		array.push( parameters.sizeAttenuation );
+		array.push( parameters.maxBones );
+		array.push( parameters.morphTargetsCount );
+		array.push( parameters.numDirLights );
+		array.push( parameters.numPointLights );
+		array.push( parameters.numSpotLights );
+		array.push( parameters.numHemiLights );
+		array.push( parameters.numRectAreaLights );
+		array.push( parameters.numDirLightShadows );
+		array.push( parameters.numPointLightShadows );
+		array.push( parameters.numSpotLightShadows );
+		array.push( parameters.shadowMapType );
+		array.push( parameters.toneMapping );
+		array.push( parameters.numClippingPlanes );
+		array.push( parameters.numClipIntersection );
+		array.push( parameters.format );
+		array.push( parameters.specularColorMapEncoding );
+		array.push( parameters.sheenColorMapEncoding );
+
+	}
+
+	function getProgramCacheKeyBooleans( array, parameters ) {
+
+		let booleancount = 0;
+		if ( parameters.isWebGL2 )
+			booleancount += 1;
+		if ( parameters.supportsVertexTextures )
+			booleancount += 2;
+		if ( parameters.instancing )
+			booleancount += 4;
+		if ( parameters.instancingColor )
+			booleancount += 8;
+		if ( parameters.map )
+			booleancount += 16;
+		if ( parameters.matcap )
+			booleancount += 32;
+		if ( parameters.envMap )
+			booleancount += 64;
+		if ( parameters.envMapCubeUV )
+			booleancount += 128;
+		if ( parameters.lightMap )
+			booleancount += 256;
+		if ( parameters.aoMap )
+			booleancount += 512;
+		if ( parameters.emissiveMap )
+			booleancount += 1024;
+		if ( parameters.bumpMap )
+			booleancount += 2048;
+		if ( parameters.normalMap )
+			booleancount += 4096;
+		if ( parameters.objectSpaceNormalMap )
+			booleancount += 8192;
+		if ( parameters.tangentSpaceNormalMap )
+			booleancount += 16384;
+		if ( parameters.clearcoat )
+			booleancount += 32768;
+		if ( parameters.clearcoatMap )
+			booleancount += 65536;
+		if ( parameters.clearcoatRoughnessMap )
+			booleancount += 131072;
+		if ( parameters.clearcoatNormalMap )
+			booleancount += 262144;
+		if ( parameters.displacementMap )
+			booleancount += 524288;
+		if ( parameters.specularMap )
+			booleancount += 1048576;
+		if ( parameters.roughnessMap )
+			booleancount += 2097152;
+		if ( parameters.metalnessMap )
+			booleancount += 4194304;
+		if ( parameters.gradientMap )
+			booleancount += 8388608;
+		if ( parameters.alphaMap )
+			booleancount += 16777216;
+		if ( parameters.alphaTest )
+			booleancount += 33554432;
+		if ( parameters.vertexColors )
+			booleancount += 67108864;
+		if ( parameters.vertexAlphas )
+			booleancount += 134217728;
+		if ( parameters.vertexUvs )
+			booleancount += 268435456;
+		if ( parameters.vertexTangents )
+			booleancount += 536870912;
+		if ( parameters.uvsVertexOnly )
+			booleancount += 1073741824;
+		if ( parameters.fog )
+			booleancount += 2147483648;
```

Comment:
How about using a bitmask via `THREE.Layers`? So create a layer object on the module scope and then just use `enable()` in the range of 1 - 31. Use `disableAll()` at the beginning and evaluate the `mask` property at the end. Something like:

At the end you do this:
```js
_programLayers.disableAll();

if ( parameters.isWebGL2 ) _programLayers.enable( 1 );
if ( parameters.supportsVertexTextures ) _programLayers.enable( 2 );
if ( parameters.instancing ) _programLayers.enable( 3 );

// ...

array.push( _programLayers.mask );

_programLayers.disableAll();

if ( parameters.useFog ) _programLayers.enable( 1 );
if ( parameters. flatShading ) _programLayers.enable( 2 );
if ( parameters. logarithmicDepthBuffer ) _programLayers.enable( 3 );

// ...

Reply:
I've added the layers code.